src/Propulsion.MessageDb/MessageDbSource.fs

```diff
@@ -0,0 +1,97 @@
+namespace Propulsion.MessageDb
+
+open System
+open System.Data.Common
+open System.Diagnostics
+open FSharp.Control
+open FsCodec
+open FsCodec.Core
+open Npgsql
+open NpgsqlTypes
+open Propulsion.Feed
+
+type MessageDbCategoryReader(connectionString) =
+    let readonly (bytes: byte array) = ReadOnlyMemory.op_Implicit(bytes)
+    let readRow (reader: DbDataReader) =
+        let readNullableString idx = if reader.IsDBNull(idx) then None else Some (reader.GetString idx)
+        let timestamp = DateTimeOffset(DateTime.SpecifyKind(reader.GetDateTime(7), DateTimeKind.Utc))
+        let streamName = reader.GetString(8)
+        let event =TimelineEvent.Create(
+           index = reader.GetInt64(0),
+           eventType = reader.GetString(1),
+           data = (reader.GetFieldValue<byte array>(2) |> readonly),
+           meta = (reader.GetFieldValue<byte array>(3) |> readonly),
+           eventId = reader.GetGuid(4),
+           ?correlationId = readNullableString 5,
+           ?causationId = readNullableString 6,
+           timestamp = timestamp)
+
+        struct(StreamName.parse streamName, event)
+    member _.ReadCategoryMessages(category: TrancheId, fromPositionInclusive: int64, batchSize: int, ct) = task {
+        use conn = new NpgsqlConnection(connectionString)
+        do! conn.OpenAsync(ct)
+        let command = conn.CreateCommand()
+        command.CommandText <- "select
+                                   global_position, type, data, metadata, id::uuid,
```

Comment:
This doesnt seem to match L20 - `Index` is the event offset in the specific stream; my guess is `global_position` is not that?

Reply:
Very true, The global_position is what we checkpoint against so I wired it through the index. Should I wire it into the `ctx` instead via a `box global_position`?

Reply:
In most other cases it seems we return a "page" that contains the checkpoint. I'll do the same here